File: dailycodingproblem/problem5/main.js
function cons(a, b) {
    function pair(f) {
        return f(a, b);
    }
    return pair;
}

function car(pair) {
    function first(a, b) {
        return a;
    }

    return pair(first);
}

function cdr(pair) {
    function last(a, b) {
        return b;
    }

    return pair(last);
}

console.log(car(cons(3, 4)) === 3);
console.log(cdr(cons(3, 4)) === 4);

File: dailycodingproblem/problem5/main.cpp
#include <cassert>
#include <iostream>
#include <functional>

int main()
{
    auto cons = [](int a, int b) {
        auto pair = [&](std::function<int(int, int)> f) {
            return f(a, b);
        };

        return pair;
    };

    auto car = [](std::function<int(std::function<int(int, int)>)> pair) -> int {
        auto first = [](int a, int b) -> int {
            return a;
        };

        return pair(first);
    };

    auto cdr = [](std::function<int(std::function<int(int, int)>)> pair) -> int {
        auto last = [](int a, int b) -> int {
            return b;
        };

        return pair(last);
    };

    assert(car(cons(3, 4)) == 3);
    assert(cdr(cons(3, 4)) == 4);
    std::cout << "All tests PASSED!" << std::endl;

    return 0;
}
